src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php

```diff
@@ -62,6 +70,25 @@ public function process(ContainerBuilder $container)
                 $tmpContainer->addObjectResource($configuration);
             }
 
+            if (null !== $envTypes && method_exists(BaseNode::class, 'setPlaceholder')) {
+                $defaultEnvBag = new ParameterBag($parameters);
+                foreach ($resolvingBag->getEnvPlaceholders() as $env => $placeholders) {
+                    $prefix = (false === $i = strpos($env, ':')) ? 'string' : substr($env, 0, $i);
+                    $types = isset($envTypes[$prefix]) ? $envTypes[$prefix] : array('string');
+                    $default = ($hasEnv = (false === $i && $defaultEnvBag->has("env($env)"))) ? $defaultEnvBag->get("env($env)") : null;
+                    if (null !== $default && !in_array($type = gettype($default), $types, true)) {
+                        throw new LogicException(sprintf('Invalid type for env parameter "env(%s)", expected "%s" got "%s".', $env, implode('", "', $types), $type));
+                    }
+                    $values = array();
+                    foreach ($types as $type) {
+                        $values[$type] = $hasEnv ? $default : (isset(self::$typeFixtures[$type]) ? self::$typeFixtures[$type] : null);
+                    }
+                    foreach ($placeholders as $placeholder) {
+                        BaseNode::setPlaceholder($placeholder, $values);
```

Comment:
We must set a known state for placeholders, and reset it after processing to avoid state leaks.